packages/caravan-fees/src/btcTransactionTemplate.ts

```diff
@@ -0,0 +1,292 @@
+  }
+
+  /**
+   * Calculates the current fee of the transaction.
+   * @returns The current fee in satoshis
+   */
+  get currentFee(): Satoshis {
+    return this.getTotalInputAmount() - this.getTotalOutputAmount();
+  }
+
+  /**
+   * Checks if the current fees are sufficient to meet the target fee rate.
+   * @returns True if the fees are paid, false otherwise
+   */
+  areFeesPayPaid(): boolean {
+    return this.currentFee >= this.targetFeesToPay;
+  }
+
+  /**
+   * Checks if the current fee rate meets or exceeds the target fee rate.
+   * @returns True if the fee rate is satisfied, false otherwise
+   */
+  get feeRateSatisfied(): boolean {
+    return this.estimatedFeeRate >= this._targetFeeRate;
+  }
+
+  /**
+   * Determines if a change output is needed.
+   * @returns True if a change output is needed, false otherwise
+   */
+  get needsChangeOutput(): boolean {
+    const targetFeesWithDustBuffer = this.targetFeesToPay + this._dustThreshold;
+    return (
+      !this.malleableOutputs.length &&
+      this.currentFee > targetFeesWithDustBuffer
+    );
+  }
+
+  /**
+   * Calculates the total input amount.
+   * @returns The total input amount in satoshis
+   */
+  getTotalInputAmount(): Satoshis {
+    return this._inputs.reduce((sum, input) => sum + input.amountSats, 0);
+  }
+
+  /**
+   * Calculates the total output amount.
+   * @returns The total output amount in satoshis
+   */
+  getTotalOutputAmount(): Satoshis {
+    return this._outputs.reduce((sum, output) => sum + output.amountSats, 0);
+  }
+
+  /**
+   * Estimates the virtual size of the transaction.
+   * @returns The estimated virtual size in vbytes
+   */
+  get estimatedVsize(): number {
+    return estimateTransactionVsize({
+      addressType: this._scriptType,
+      numInputs: this._inputs.length,
+      numOutputs: this._outputs.length,
+      m: this._requiredSigners,
+      n: this._totalSigners,
+    });
+  }
+
+  /**
+   * Calculates the estimated fee rate of the transaction.
+   * @returns The estimated fee rate in satoshis per vbyte
+   */
+  get estimatedFeeRate(): number {
+    return this.currentFee / this.estimatedVsize;
+  }
+
+  /**
+   * Adds an input to the transaction.
+   * @param input - The input to add
+   */
+  addInput(input: BtcTxInputTemplate): void {
+    this._inputs.push(input);
+  }
+
+  /**
+   * Adds an output to the transaction.
+   * @param output - The output to add
+   */
+  addOutput(output: BtcTxOutputTemplate): void {
+    this._outputs.push(output);
+  }
+
+  /**
+   * Removes an output from the transaction.
+   * @param index - The index of the output to remove
+   */
+  removeOutput(index: number): void {
+    this._outputs.splice(index, 1);
+  }
+
+  /**
+   * Adjusts the change output of the transaction.
+   * If the new change amount is less than the dust threshold, the change output is removed.
+   */
+  adjustChangeOutput(): void {
+    const changeOutput = this.malleableOutputs[0];
+    if (!changeOutput) return;
```

Comment:
[fixed](https://github.com/caravan-bitcoin/caravan/pull/114/commits/3c9f3ffa10cc1c7c57c8482c976adec5678805c1#diff-c113a33821d4e86e4ad4f20743227a51337831792cc5c5926aa2a5971c11d7b1R293)

also changed the logic to filter from change outputs ... rather than all malleable ones